src/dbnode/storage/namespace.go

```diff
@@ -812,63 +835,50 @@ func (n *dbNamespace) Bootstrap(start time.Time, process bootstrap.Process) erro
 		return nil
 	}
 
-	var (
-		owned  = n.GetOwnedShards()
-		shards = make([]databaseShard, 0, len(owned))
-	)
-	for _, shard := range owned {
-		if !shard.IsBootstrapped() {
-			shards = append(shards, shard)
-		}
-	}
-	if len(shards) == 0 {
-		success = true
-		n.metrics.bootstrap.ReportSuccess(n.nowFn().Sub(callStart))
-		return nil
-	}
-
-	shardIDs := make([]uint32, len(shards))
-	for i, shard := range shards {
-		shardIDs[i] = shard.ID()
-	}
-
-	bootstrapResult, err := process.Run(start, metadata, shardIDs)
-	if err != nil {
-		n.log.Error("bootstrap aborted due to error",
-			zap.Stringer("namespace", n.id),
-			zap.Error(err))
-		return err
-	}
-	n.metrics.bootstrap.Success.Inc(1)
-
 	// Bootstrap shards using at least half the CPUs available
 	workers := xsync.NewWorkerPool(int(math.Ceil(float64(runtime.NumCPU()) / 2)))
 	workers.Init()
 
-	numSeries := bootstrapResult.DataResult.ShardResults().NumSeries()
-	n.log.Info("bootstrap data fetched now initializing shards with series blocks",
-		zap.Int("numShards", len(shards)),
-		zap.Int64("numSeries", numSeries),
-	)
-
 	var (
-		multiErr = xerrors.NewMultiError()
-		results  = bootstrapResult.DataResult.ShardResults()
-		mutex    sync.Mutex
-		wg       sync.WaitGroup
+		bootstrappedShards = bootstrapResult.Shards
+		multiErr           = xerrors.NewMultiError()
+		mutex              sync.Mutex
+		wg                 sync.WaitGroup
 	)
-	for _, shard := range shards {
-		shard := shard
-		wg.Add(1)
-		workers.Go(func() {
-			var bootstrapped *result.Map
-			if shardResult, ok := results[shard.ID()]; ok {
-				bootstrapped = shardResult.AllSeries()
-			} else {
-				bootstrapped = result.NewMap(result.MapOptions{})
+	n.log.Info("bootstrap marking all shards as bootstrapped",
+		zap.Stringer("namespace", n.id),
+		zap.Int("numShards", len(bootstrappedShards)),
+		zap.Int("dataNumSeries", bootstrapResult.DataMetadata.NumSeries),
+		zap.Int("indexNumSeries", bootstrapResult.IndexMetadata.NumSeries))
+	for _, shard := range n.GetOwnedShards() {
+		// Make sure it was bootstrapped during this bootstrap run.
+		shardID := shard.ID()
+		bootstrapped := false
+		for _, elem := range bootstrappedShards {
+			if elem == shardID {
+				bootstrapped = true
+				break
 			}
+		}
+		if !bootstrapped {
+			// NB(r): Not bootstrapped in this bootstrap run.
+			continue
+		}
 
-			err := shard.Bootstrap(bootstrapped)
+		if shard.IsBootstrapped() {
+			// Since no concurrent bootstraps, we should never
+			// have just bootstrapped this shard if it's not already
+			// bootstrapped.
```

Comment:
This comment is a little confusing; does it mean if the shard is already marked as bootstrapped, then there can never be a bootstrap update to this shard?